tools/octree_viewer.cpp

```diff
@@ -334,26 +373,62 @@ class OctreeViewer
   void extractPointsAtLevel(int depth)
   {
     displayCloud->points.clear();
+    cloudVoxel->points.clear();
 
     pcl::octree::OctreePointCloudVoxelCentroid<pcl::PointXYZ>::Iterator tree_it;
     pcl::octree::OctreePointCloudVoxelCentroid<pcl::PointXYZ>::Iterator tree_it_end = octree.end();
 
-    pcl::PointXYZ pt;
+    pcl::PointXYZ pt_voxel_center;
+    pcl::PointXYZ pt_centroid;
     std::cout << "===== Extracting data at depth " << depth << "... " << std::flush;
     double start = pcl::getTime ();
 
     for (tree_it = octree.begin(depth); tree_it!=tree_it_end; ++tree_it)
     {
+      // If the iterator is not at the right depth, continue
       if (tree_it.getCurrentOctreeDepth () != depth)
         continue;
 
+      // Compute the point at the center of the voxel which represents the current OctreeNode
       Eigen::Vector3f voxel_min, voxel_max;
-      octree.getVoxelBounds(tree_it, voxel_min, voxel_max);
+      octree.getVoxelBounds (tree_it, voxel_min, voxel_max);
+
+      pt_voxel_center.x = (voxel_min.x () + voxel_max.x ()) / 2.0f;
+      pt_voxel_center.y = (voxel_min.y () + voxel_max.y ()) / 2.0f;
+      pt_voxel_center.z = (voxel_min.z () + voxel_max.z ()) / 2.0f;
+      cloudVoxel->points.push_back (pt_voxel_center);
+
+      // If the asked depth is the depth of the octree, retrieve the centroid at this LeafNode
+      if ( octree.getTreeDepth() == depth ){
+        pcl::octree::OctreePointCloudVoxelCentroid<pcl::PointXYZ>::LeafNode* container = static_cast<pcl::octree::OctreePointCloudVoxelCentroid<pcl::PointXYZ>::LeafNode*> (tree_it.getCurrentOctreeNode ());
+
+        container->getContainer().getCentroid (pt_centroid);
+      }
+      // Else, compute the centroid of the LeafNode under the current BranchNode
+      else {
+        // Reset the current centroid coordinates
+        pt_centroid = pcl::PointXYZ(0, 0, 0);
+
+        // Retrieve every centroid under the current BranchNode
+        pcl::octree::OctreeKey dummy_key;
+        pcl::PointCloud<pcl::PointXYZ>::VectorType voxelCentroids;
+        octree.getVoxelCentroidsRecursive (static_cast<pcl::octree::OctreePointCloudVoxelCentroid<pcl::PointXYZ>::BranchNode*> (*tree_it), dummy_key, voxelCentroids);
+
+        // Iterate over the leafs to compute the centroid of all of them
```

Comment:
We have a dedicated utility class for this: [`CentroidPoint`](http://docs.pointclouds.org/trunk/classpcl_1_1_centroid_point.html#ae98c53f8bbd42f13660d198ad3855ceb).

```cpp
// Add this include
#include <pcl/common/centroid.h>

// Utility class, instantiate before the loop
CentroidPoint<pcl::PointXYZ> centroid;

// Inside the loop add points
centroid.add (voxelCentroids[j]);

// Fetch centroid after the loop
centroid.get (pt_centroid);
```

Note that the line 410 is not necessary.